crates/polars-ops/src/series/ops/index_of.rs

```diff
@@ -0,0 +1,137 @@
+                    return Some(index);
+                } else {
+                    index += 1;
+                }
+            }
+        }
+    }
+    None
+}
+
+impl<'a, T> ChunkSearch<'a, T::Native> for ChunkedArray<T>
+where
+    T: PolarsNumericType,
+{
+    fn index_of(&'a self, value: Option<T::Native>) -> Option<usize> {
+        // A NaN is never equal to anything, including itself. But we still want
+        // to be able to search for NaNs, so we handle them specially.
+        if value.map(|v| v.is_nan()) == Some(true) {
+            return index_of_predicate(self, |v| v.is_nan());
+        }
+
+        // Searching for an actual value:
+        if let Some(value) = value {
+            return index_of_predicate(self, |v| *v == value);
+        }
+
+        // Searching for null:
+        let mut index = 0;
+        for chunk in self.chunks() {
+            let length = chunk.len();
+            if let Some(bitmap) = chunk.validity() {
+                let leading_ones = bitmap.leading_ones();
+                if leading_ones < length {
+                    return Some(index + leading_ones);
+                }
+            } else {
+                index += length;
+            }
+        }
+        None
+    }
+}
+
+/// Try casting the value to the correct type, then call index_of().
+macro_rules! try_index_of_numeric_ca {
+    ($ca:expr, $value:expr) => {{
+        let ca = $ca;
+        let value = $value;
+        if value == AnyValue::Null {
+            return Ok(ca.index_of(None));
+        }
+        let cast_value = cast_if_lossless(&value, ca.dtype());
+        if cast_value.is_some() {
+            ca.index_of(cast_value.map(|v| v.extract().unwrap()))
+        } else {
+            // We can't cast the searched-for value to a valid data point within
+            // the dtype of the Series we're searching, which means we will
+            // never find that value.
+            None
+        }
+    }};
+}
+
+/// Cast to the dtype, but return ``None`` if the casting changed the value.
+pub fn cast_if_lossless<'a>(value: &'a AnyValue, dtype: &'a DataType) -> Option<AnyValue<'a>> {
+    let result = value.cast(dtype);
+    let value_dtype = value.dtype();
+    let roundtrip = result.cast(&value_dtype);
+    if &roundtrip == value {
+        Some(result)
+    } else {
+        None
+    }
+}
+
+/// Find the index of a given value (the first and only entry in `value_series`)
+/// within the series.
+pub fn index_of(series: &Series, value: &AnyValue<'_>) -> PolarsResult<Option<usize>> {
+    let value_dtype = value.dtype();
+
+    let value = match value_dtype {
+        dtype if dtype.is_signed_integer() => value.cast(&DataType::Int64),
+        dtype if dtype.is_unsigned_integer() => value.cast(&DataType::UInt64),
+        dtype if dtype.is_float() => value.cast(&DataType::Float64),
+        DataType::Null => AnyValue::Null,
+        _ => unimplemented!("index_of() not yet supported for dtype {:?}", value_dtype),
```

Comment:
The other types (except boolean) can go into the row-encoding and use the branch for binary data.

Reply:
I think it should be in one go, as I think it can all go to a single generic impl.

Reply:
The row-encoding is used in multi-key group-bys/joins and joins on nested types. I would recommend reading in the join dispatch. Can give a bit more pointers, if you get lost. :)

Reply:
OK will go read.